gettext-rs/src/macros.rs

```diff
@@ -1,50 +1,97 @@
+/// This is an implementation detail for counting arguments in the gettext macros. Don't call this directly.
+#[macro_export]
+#[doc(hidden)]
+macro_rules! count_args {
+    () => { 0 };
+    ($_e: expr $(, $rest: expr)*) => { 1 + $crate::count_args!($($rest),*) }
+}
+
+/// This is an implementation detail for replacing arguments in the gettext macros. Don't call this directly.
+#[macro_export]
+#[doc(hidden)]
+macro_rules! freplace {
+    ($input:expr, $($args:expr),+ $(,)?) => {{
+        let mut parts = $input.split("{}");
+        debug_assert_eq!(parts.clone().count() - 1, $crate::count_args!($($args),*), "Argument count has to match number of format directivies ({{}})");
+
+        let mut output = parts.next().unwrap_or_default().to_string();
+        $(
+            output += &format!("{}{}", $args, &parts.next().expect("Argument count has to match number of format directivies ({})"));
+        )*
+        output
+    }};
+}
 
 /// Like [gettext](fn.gettext.html), but allows for formatting.
 #[macro_export]
 macro_rules! gettext {
-    ($($arg:tt)*) => ($crate::gettext(std::fmt::format(format_args!($($arg)*))))
+    ($format:expr, $($args:expr),+ $(,)?) => {{
+        let s = $crate::gettext($format);
+        $crate::freplace!(s, $($args),*)
+    }};
 }
 
 /// Like [dgettext](fn.dgettext.html), but allows for formatting.
 #[macro_export]
 macro_rules! dgettext {
-    ($domain:expr, $string:expr, $($string_names:ident = $string_values:expr),*) => ($crate::dgettext($domain, format!($string, $( $string_names = $string_values ),*)));
+    ($domain:expr, $format:expr, $($args:expr),+ $(,)?) => {{
+        let s = $crate::dgettext($domain, $format);
+        $crate::freplace!(s, $($args),*)
+    }};
 }
 
-/// Like [dcgettext](fn.gettext.html), but allows for formatting.
+/// Like [dcgettext](fn.dcgettext.html), but allows for formatting.
 #[macro_export]
 macro_rules! dcgettext {
-    ($domain:expr, $category:expr, $string:expr, $($string_names:ident = $string_values:expr),*,) => ($crate::dcgettext($domain, format!($string, $( $string_names = $string_values ),*), $category));
+    ($domain:expr, $category:expr, $format:expr, $($args:expr),+ $(,)?) => {{
+        let s = $crate::dcgettext($domain, $format, $category);
+        $crate::freplace!(s, $($args),*)
+    }};
 }
 
-/// Like [ngettext](fn.gettext.html), but allows for formatting.
+/// Like [ngettext](fn.ngettext.html), but allows for formatting.
 #[macro_export]
 macro_rules! ngettext {
-    ($singular:expr, $($singular_names:ident = $singular_values:expr),*, $plural:expr, $($plural_names:ident = $plural_values:expr),*, $n:expr) => ($crate::ngettext(format!($singular, $( $singular_names = $singular_values ),*), format!($plural, $( $plural_names = $plural_values ),*), $n));
+    ($singular:expr, $multiple:expr, $n:expr, $($args:expr),+ $(,)?) => {{
```

Comment:
Plural case is "plural", not "multiple" ;) Why change the name?

(Ditto in other macros below.)